src/client/Microsoft.Identity.Client/ManagedIdentity/V2/MtlsBindingCache.cs

```diff
@@ -0,0 +1,95 @@
+        private readonly IPersistentCertificateCache _persisted;
+
+        /// <summary>
+        /// Inject both caches to avoid global state and enable testing.
+        /// </summary>
+        public MtlsBindingCache(ICertificateCache memory, IPersistentCertificateCache persisted)
+        {
+            _memory = memory ?? throw new ArgumentNullException(nameof(memory));
+            _persisted = persisted ?? throw new ArgumentNullException(nameof(persisted));
+        }
+
+        public async Task<Tuple<X509Certificate2, string, string>> GetOrCreateAsync(
+            string cacheKey,
+            Func<Task<Tuple<X509Certificate2, string, string>>> factory,
+            CancellationToken cancellationToken,
+            ILoggerAdapter logger)
+        {
+            if (string.IsNullOrWhiteSpace(cacheKey))
+                throw new ArgumentException("cacheKey must be non-empty.", nameof(cacheKey));
+            if (factory is null)
+                throw new ArgumentNullException(nameof(factory));
+
+            // 1) In-memory cache first
+            if (_memory.TryGet(cacheKey, out var cached, logger))
+            {
+                return Tuple.Create(cached.Certificate, cached.Endpoint, cached.ClientId);
+            }
+
+            // 2) Per-key gate (dedupe concurrent mint)
+            await _gates.EnterAsync(cacheKey, cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                // Re-check after acquiring the gate
+                if (_memory.TryGet(cacheKey, out cached, logger))
+                {
+                    return Tuple.Create(cached.Certificate, cached.Endpoint, cached.ClientId);
+                }
+
+                // 3) Persistent cache (best-effort)
+                if (_persisted.Read(cacheKey, out var persisted, logger))
+                {
+                    if (persisted.Certificate.HasPrivateKey)
+                    {
+                        var v = new CertificateCacheValue(persisted.Certificate, persisted.Endpoint, persisted.ClientId);
+                        _memory.Set(cacheKey, in v, logger);
+                        return Tuple.Create(v.Certificate, v.Endpoint, v.ClientId);
+                    }
+
+                    // Defensive: persisted entry is unusable; dispose and mint new
+                    persisted.Certificate.Dispose();
+                    logger?.Verbose(() => "[PersistentCert] Skipping persisted cert without private key; minting new.");
+                }
+
+                // 4) Mint + back-fill mem + best-effort persist + prune
+                var created = await factory().ConfigureAwait(false);
+                var createdValue = new CertificateCacheValue(created.Item1, created.Item2, created.Item3);
+
+                _memory.Set(cacheKey, in createdValue, logger);
+                _persisted.Write(cacheKey, created.Item1, created.Item2, logger);
+                _persisted.Delete(cacheKey, logger);
```

Comment:
can you add a comment here? Why are you deleting right after you write?

Reply:
Write persists the newest binding for a given alias, and Delete is used to prune expired entries for that alias only, so the store doesn't grow unbounded over time. Added code comments to clarify